### PR TITLE
Add support for conf.d configuration files

### LINKED ISSUE #148: Add support for conf.d configuration files
This is a follow-up of https://github.com/containers/hirte/issues/147 where it's suggested to use `/etc/hirte/hirte.conf` and `/etc/hirte/agent.conf` as configuration files, which could be modified by users to specify custom options.

But for automation it's not easy to parse current configuration file and save a new one with modified values, it would be much easier to use `conf.d` directory with custom user configuration files.

Here is the suggestion for the whole configuration files lineup for `hirte` manager:

* `/usr/share/hirte/config/hirte-default.conf`
   * The base configuration file provided project defaults, which should not be modified by users
* `/etc/hirte/hirte.conf`
   * The configurations file, where users can modified default values 
* `/etc/hirte/hirte.conf.d/`
   * Configuration directory where users might insert a configuration file (for example `99-custom-logging.conf`)
   * Only files with `.conf` suffix are used to load configuration changes
   * Files in this directory are loaded in alphabetical order, so it's suggested to use number prefix to ease understanding of loading (for example `95-my-manager.conf` and `99-custom-logging.conf`)
  
Similar structure should be used also for `hirte-agent`:

* `/usr/share/hirte-agent/config/hirte-default.conf`
* `/etc/hirte/agent.conf`
* `/etc/hirte/agent.conf.d/`

Using above configuration files structure should also change the order of loading configurations files during `hirte`/`hirte-agent` startup:

1. Load the default configuration values from `/usr/share` and store them to the configuration hash map
2. Load user customizations from `/etc/hirte/{hirte|agent}.conf` and overwrite updated values in the configuration hash map
3. Load user customizations files from `/etc/hirte/{hirte|agent}.conf.d/` by alphabetical order and overwrite updated values in the configuration hash map
4. Check if all required options have a valid value

### PR DESCRIPTION
**Pull request by ArtiomDivak**
This is a follow-up of #147 where it's suggested to use /etc/hirte/hirte.conf and /etc/hirte/agent.conf as configuration files, which could be modified by users to specify custom options.

But for automation it's not easy to parse current configuration file and save a new one with modified values, it would be much easier to use conf.d directory with custom user configuration files.

Here is the suggestion for the whole configuration files lineup for hirte manager:

    /usr/share/hirte/config/hirte-default.conf
        The base configuration file provided project defaults, which should not be modified by users
    /etc/hirte/hirte.conf
        The configurations file, where users can modified default values
    /etc/hirte/hirte.conf.d/
        Configuration directory where users might insert a configuration file (for example 99-custom-logging.conf)
        Only files with .conf suffix are used to load configuration changes
        Files in this directory are loaded in alphabetical order, so it's suggested to use number prefix to ease understanding of loading (for example 95-my-manager.conf and 99-custom-logging.conf)

Similar structure should be used also for hirte-agent:

    /usr/share/hirte-agent/config/hirte-default.conf
    /etc/hirte/agent.conf
    /etc/hirte/agent.conf.d/

Using above configuration files structure should also change the order of loading configurations files during hirte/hirte-agent startup:

    1)Load the default configuration values from /usr/share and store them to the configuration hash map
    2)Load user customizations from /etc/hirte/{hirte|agent}.conf and overwrite updated values in the configuration hash map
    3)Load user customizations files from /etc/hirte/{hirte|agent}.conf.d/ by alphabetical order and overwrite updated values in the configuration hash map
    4)Check if all required options have a valid value

Fixes: #148
Signed-off-by: Artiom Divak adivak@redhat.com

### CODE REVIEW

File: config/agent/README-agent [hunk 1]
```diff
@@ -0,0 +1,3 @@
+This directory is for user costume configuration for the agent.
```
Comment (mwperina) [inline config/agent/README-agent]: I would name this file just `README.md` and put it under `config/agent/agent.conf.d` subdirectory

And I'd rephrase as below:
```
This directory is used for user configuration of `hirte-agent`. Each configuration file should end with `.conf` suffix, otherwise it's ignored. Configuration files from this directory are loaded by alphabetical order, configuration from a file loaded later overwrites the previous configuration.

It's suggested to name configuration files using `<NN>-<FILE NAME>.conf` template, where `<NN>` is a number, for example `95-debug-logging.conf`.

File: config/hirte/README-hirte [hunk 2]
```diff
@@ -0,0 +1,3 @@
+This directory is for user costume configuration for hirte.
```
Comment (mwperina) [inline config/hirte/README-hirte]: Please rename the file to `README.md`, put it under `config/hirte/hirte.conf.d` subdirectory and use similar content as for the agent

File: config/meson.build [hunk 3]
```diff
@@ -19,3 +19,13 @@ install_data(
   'hirte/hirte-default.conf',
   install_dir : join_paths(get_option('datadir'), 'hirte', 'config')
 )
+
+install_data(
+  'agent/README-agent',
+  install_dir :  get_option('sysconfdir') / 'hirte' / 'agent.conf.d'
```
Comment (mwperina) [inline config/meson.build]: Please use join_paths() function

File: config/meson.build [hunk 4]
```diff
@@ -19,3 +19,13 @@ install_data(
...
+)
+
+install_data(
+  'hirte/README-hirte',
+  install_dir :  get_option('sysconfdir') / 'hirte' / 'hirte.conf.d'
```
Comment (mwperina) [inline config/meson.build]: Please use join_paths() function

File: hirte.spec.in [hunk 5]
```diff
@@ -83,6 +83,7 @@ This package contains the node agent.
 %{_mandir}/man1/hirtectl.*
 %{_mandir}/man5/hirte.conf.*
 %{_sysconfdir}/dbus-1/system.d/org.containers.hirte.conf
```
Comment (mwperina) [inline hirte.spec.in]: Please add following lines after line 71:

```
%dir %{_sysconfdir}/hirte
%dir %{_sysconfdir}/hirte/hirte.conf.d
```

File: hirte.spec.in [hunk 6]
```diff
@@ -83,6 +83,7 @@ This package contains the node agent.
...
+%{_sysconfdir}/hirte/hirte.conf.d/README-hirte
```
Comment (mwperina) [inline hirte.spec.in]: Please change to

```
%{_sysconfdir}/hirte/hirte.conf.d/README.md
```

File: hirte.spec.in [hunk 7]
```diff
@@ -95,6 +96,7 @@ This package contains the node agent.
 %{_mandir}/man1/hirte-agent.*
```
Comment (mwperina) [inline hirte.spec.in]: Please add following lines after line 91:

%dir %{_sysconfdir}/hirte
%dir %{_sysconfdir}/hirte/agent.conf.d

File: hirte.spec.in [hunk 8]
```diff
@@ -95,6 +96,7 @@ This package contains the node agent.
...
 %{_mandir}/man5/hirte-agent.conf.*
 %{_sysconfdir}/dbus-1/system.d/org.containers.hirte.Agent.conf
+%{_sysconfdir}/hirte/hirte.conf.d/README-agent
```
Comment (mwperina) [inline hirte.spec.in]: Please change to
```
%{_sysconfdir}/hirte/agent.conf.d/README.md
```

File: src/libhirte/common/cfg.c [hunk 9]
```diff
@@ -133,6 +138,96 @@ int cfg_load_from_file(struct config *config, const char *config_file) {
         return ini_parse(config_file, parsing_handler, config);
 }
 
+void add_file_to_list_and_sort(char *file_location) {
```
Comment (mwperina) [inline src/libhirte/common/cfg.c]: Wouldn't it be better to use list macros from https://github.com/containers/hirte/blob/main/src/libhirte/common/list.h ?

File: src/libhirte/common/cfg.c [hunk 10]
```diff
@@ -133,6 +138,96 @@ int cfg_load_from_file(struct config *config, const char *config_file) {
...
+
+void cfg_load_files_from_list(struct config *config) {
+        struct node *tmp = list_head;
+
+        if (list_head == NULL) { // nothing to load
+                return;
+        }
+
+        while (tmp != NULL) {
+                cfg_load_from_file(config, tmp->file_location);
```
Comment (mwperina) [inline src/libhirte/common/cfg.c]: Please handle errors here and return proper error code

File: src/libhirte/common/cfg.c [hunk 11]
```diff
@@ -133,6 +138,96 @@ int cfg_load_from_file(struct config *config, const char *config_file) {
...
+
+int cfg_load_from_dir(struct config *config, const char *custom_config_directory) {
+        list_head = NULL;                 // start new head
+        char *conf_file_enfing = ".conf"; // needed ending for configuration files
+        DIR *dir = opendir(custom_config_directory);
+        if (!dir) {
+                return ENOENT;
+        }
+        struct dirent *current_file = NULL;
+        while ((current_file = readdir(dir)) != NULL) { // while there is files
```
Comment (mwperina) [inline src/libhirte/common/cfg.c]: It would probably be easier to use scandir function, which could sort filename already, examples below

https://lloydrochester.com/post/c/list-directory/
https://stackoverflow.com/questions/18402428/how-to-properly-use-scandir-in-c

**Review by mwperina — CHANGES_REQUESTED** [review]
I suggest to make packaging correctly first and then perform changes around scandir and list macros

File: hirte.spec.in (hunk 5 above)
Comment (ArtiomDivak) [reply hirte.spec.in]: done

File: hirte.spec.in (hunk 7 above)
Comment (ArtiomDivak) [reply hirte.spec.in]: done

File: config/agent/README-agent.md [hunk 12]
```diff
@@ -0,0 +1,3 @@
+This directory is used for user configuration of `hirte-agent`. Each configuration file should end with `.conf` suffix, otherwise it's ignored. Configuration files from this directory are loaded by alphabetical order, configuration from a file loaded later overwrites the previous configuration.
```
Comment (mwperina) [inline config/agent/README-agent.md]: Please rename to just `README.md`, it's already distinguished by directory name

File: config/hirte/README-hirte.md [hunk 13]
```diff
@@ -0,0 +1,3 @@
+This directory is used for user configuration of `hirte-agent`. Each configuration file should end with `.conf` suffix, otherwise it's ignored. Configuration files from this directory are loaded by alphabetical order, configuration from a file loaded later overwrites the previous configuration.
```
Comment (mwperina) [inline config/hirte/README-hirte.md]: Please rename to just `README.md`, it's already distinguished by directory name

File: hirte.spec.in (hunk 5 above)
Comment (mwperina) [reply hirte.spec.in]: Not correct, you've made some errors during rebase, please check your changes with `hirte.spec.in` from the main branch

File: hirte.spec.in (hunk 7 above)
Comment (mwperina) [reply hirte.spec.in]: Same as above

**Review by mwperina — CHANGES_REQUESTED** [review]

File: config/agent/agent.conf.d/README.md [hunk 14]
```diff
@@ -0,0 +1,3 @@
+This directory is used for user configuration of `hirte-agent`. Each configuration file should end with `.conf` suffix, otherwise it's ignored. Configuration files from this directory are loaded by alphabetical order, configuration from a file loaded later overwrites the previous configuration.
```
Comment (mwperina) [inline config/agent/agent.conf.d/README.md]: The path should be only `config/agent.conf.d/README.md`

File: config/hirte/hirte.conf.d/README.md [hunk 15]
```diff
@@ -0,0 +1,3 @@
+This directory is used for user configuration of `hirte`. Each configuration file should end with `.conf` suffix, otherwise it's ignored. Configuration files from this directory are loaded by alphabetical order, configuration from a file loaded later overwrites the previous configuration.
```
Comment (mwperina) [inline config/hirte/hirte.conf.d/README.md]: The path should be only config/hirte.conf.d/README.md

File: hirte.spec.in [hunk 16]
```diff
@@ -33,6 +33,8 @@ This package contains the controller and command line tool.
 %config(noreplace) %{_sysconfdir}/hirte/hirte.conf
 %doc README.md
 %doc README.developer.md
+%dir %{_sysconfdir}/hirte
+%dir %{_sysconfdir}/hirte/hirte.conf.d
```
Comment (mwperina) [inline hirte.spec.in:35]: Please align with alphabetical order, `%dir` should land before `%doc`

File: hirte.spec.in [hunk 17]
```diff
@@ -69,6 +72,8 @@ This package contains the node agent.
 %files agent
 %config(noreplace) %{_sysconfdir}/hirte/agent.conf
 %doc README.md
+%dir %{_sysconfdir}/hirte
+%dir %{_sysconfdir}/hirte/agent.conf.d
```
Comment (mwperina) [inline hirte.spec.in:75]: Same as above

**Review by mwperina — CHANGES_REQUESTED** [review]

File: config/agent.conf.d/README.md [hunk 18]
```diff
@@ -0,0 +1,9 @@
+# Agent.conf.d
```
Comment (mwperina) [inline config/agent.conf.d/README.md]: Please change to:

```
# hirte-agent configuration directory
```

File: config/hirte.conf.d/README.md [hunk 19]
```diff
@@ -0,0 +1,9 @@
+# Hirte.conf.d
```
Comment (mwperina) [inline config/hirte.conf.d/README.md]: Please change to

```
# hirte configuration directory
```

**Review by mwperina — CHANGES_REQUESTED** [review]

**Review by mwperina — APPROVED** [review]
+1

File: src/libhirte/common/cfg.c [hunk 20]
```diff
@@ -137,6 +141,29 @@ int cfg_load_from_file(struct config *config, const char *config_file) {
         return ini_parse(config_file, parsing_handler, config);
 }
 
+char *is_file_name_ending_with_conf(char *file_name, char *conf_file_enfing) {
```
Comment (Yarboa) [inline src/libhirte/common/cfg.c]: is it a type? *conf_file_enfing should it be *conf_file_ending

File: src/agent/agent.c [hunk 21]
```diff
@@ -465,9 +465,10 @@ bool agent_parse_config(Agent *agent, const char *configfile) {
                         agent->config,
                         CFG_AGENT_DEFAULT_CONFIG,
                         CFG_ETC_HIRTE_AGENT_CONF,
-                        NULL); // TODO: https://github.com/containers/hirte/issues/148
-        if (result != 0 && result != -ENOENT) {
-                fprintf(stderr, "Error loading configuration: '%s'.\n", strerror(-result));
+                        CFG_ETC_AGENT_CONF_DIR);
+        if (result != 0) {
+                fprintf(stderr, "Error loading configuration: '%s'.", strerror(result));
```
Comment (engelmi) [inline src/agent/agent.c]:
```suggestion
                fprintf(stderr, "Error loading configuration: '%s'.", strerror(-result));
```
We are returning error codes such as `-ENOENT`. Note the negation in front, which is used to identify errors. For `strerror` to find the correct description for it, we need to revert the negation again - `-result` in this case.

File: src/libhirte/common/cfg.c [hunk 22]
```diff
@@ -137,6 +141,29 @@ int cfg_load_from_file(struct config *config, const char *config_file) {
...
+
+int cfg_load_from_dir(struct config *config, const char *custom_config_directory) {
+        struct dirent **all_files_in_dir = NULL;
+        int number_of_files = 0;
+        char *conf_file_enfing = ".conf";
+
+        number_of_files = scandir(custom_config_directory, &all_files_in_dir, NULL, alphasort);
+        if (number_of_files == -1) {
+                return ENOENT;
+        }
```
Comment (engelmi) [inline src/libhirte/common/cfg.c]:
```suggestion
        number_of_files = scandir(custom_config_directory, &all_files_in_dir, NULL, alphasort);
        if (number_of_files < 0) {
                return -errno;
        }
```
According to man page (see [here](https://man7.org/linux/man-pages/man3/scandir.3.html)), if `scandir` fails, `errno` will be set - which then contains the last error that occurred. Since `scandir` could also fail due to other reasons (e.g. `ENOTDIR`), we should return it.

File: src/libhirte/common/cfg.c [hunk 23]
```diff
@@ -137,6 +141,29 @@ int cfg_load_from_file(struct config *config, const char *config_file) {
         return ini_parse(config_file, parsing_handler, config);
 }
 
+char *is_file_name_ending_with_conf(char *file_name, char *conf_file_enfing) {
+        return strstr(file_name, conf_file_enfing);
+}
+
+int cfg_load_from_dir(struct config *config, const char *custom_config_directory) {
+        struct dirent **all_files_in_dir = NULL;
```
Comment (engelmi) [inline src/libhirte/common/cfg.c:152]: I haven't run it in valgrind yet, but I think you have to free `all_files_in_dir` as `scandir` calls `malloc` to allocate the memory for the items.

File: src/libhirte/common/cfg.c [hunk 24]
```diff
@@ -137,6 +141,29 @@ int cfg_load_from_file(struct config *config, const char *config_file) {
...
+        int number_of_files = 0;
+        char *conf_file_enfing = ".conf";
+
+        number_of_files = scandir(custom_config_directory, &all_files_in_dir, NULL, alphasort);
+        if (number_of_files == -1) {
+                return ENOENT;
+        }
+        for (int i = 1; i <= number_of_files; i++) {
+                char *file_name = all_files_in_dir[i]->d_name;
+                char *files_ends_with_conf = is_file_name_ending_with_conf(file_name, conf_file_enfing);
```
Comment (engelmi) [inline src/libhirte/common/cfg.c]: If a function has the prefix `is_` in its name, I'd expect it to return a `bool`. In this case, the resulting code would be a bit more concise as well:
```C
if (is_file_name_ending_with_conf(file_name, conf_file_ending)){
   cfg_load_from_file(config, file_name);
}
```
Could you change the `is_file_name_ending_with_conf` to return a bool?

File: src/libhirte/common/cfg.c [hunk 25]
```diff
@@ -137,6 +141,29 @@ int cfg_load_from_file(struct config *config, const char *config_file) {
...
+int cfg_load_from_dir(struct config *config, const char *custom_config_directory) {
+        struct dirent **all_files_in_dir = NULL;
+        int number_of_files = 0;
+        char *conf_file_enfing = ".conf";
+
+        number_of_files = scandir(custom_config_directory, &all_files_in_dir, NULL, alphasort);
+        if (number_of_files == -1) {
+                return ENOENT;
+        }
+        for (int i = 1; i <= number_of_files; i++) {
```
Comment (engelmi) [inline src/libhirte/common/cfg.c]: Is there a specific reason to start at 1 instead of 0?

File: src/libhirte/common/cfg.c (hunk 20 above)
Comment (mwperina) [reply src/libhirte/common/cfg.c]: Or maybe just `file_suffix`?

File: src/libhirte/common/cfg.c (hunk 24 above)
Comment (mwperina) [inline src/libhirte/common/cfg.c]: Hmm, any reason why are we not using filter function in scandir instead of filterning all files returned from scandir?

**Review by mwperina — CHANGES_REQUESTED** [review]

File: src/libhirte/common/cfg.c (hunk 20 above)
Comment (ArtiomDivak) [reply src/libhirte/common/cfg.c]: Changed  to file_suffix

File: src/agent/agent.c (hunk 21 above)
Comment (ArtiomDivak) [reply src/agent/agent.c]: Added -result

File: src/libhirte/common/cfg.c (hunk 23 above)
Comment (ArtiomDivak) [reply src/libhirte/common/cfg.c:152]: Added free function for all_files_in_dir veriable

File: src/libhirte/common/cfg.c (hunk 25 above)
Comment (ArtiomDivak) [reply src/libhirte/common/cfg.c]: Changed to 0

File: src/libhirte/common/cfg.c (hunk 24 above)
Comment (ArtiomDivak) [reply src/libhirte/common/cfg.c]: Now it should be int so I could use it in scandir

Comment (ArtiomDivak) [reply src/libhirte/common/cfg.c]: Changed the is_file_name_ending_with_conf to return int and added it to the scandir

File: src/libhirte/common/cfg.c [hunk 26]
```diff
@@ -137,6 +141,33 @@ int cfg_load_from_file(struct config *config, const char *config_file) {
...
+
+        number_of_files = scandir(
+                        custom_config_directory, &all_files_in_dir, &is_file_name_ending_with_conf, alphasort);
+        if (number_of_files < 0) {
+                return -errno;
+        }
+
+        for (int i = 0; i < number_of_files; i++) {
+                char *file_name = all_files_in_dir[i]->d_name;
+                printf("%s\n", file_name);
```
Comment (mwperina) [inline src/libhirte/common/cfg.c]: Please remove this debug output

**Review by mwperina — APPROVED** [review]

**Comment by mwperina** [comment]
Please rebase

**Review by Yarboa — APPROVED** [review]
LGTM

File: src/libhirte/common/cfg.c [hunk 27]
```diff
@@ -137,6 +141,32 @@ int cfg_load_from_file(struct config *config, const char *config_file) {
         return ini_parse(config_file, parsing_handler, config);
 }
 
+int is_file_name_ending_with_conf(const struct dirent *entry) {
+        if (strstr(entry->d_name, ".conf") != NULL) {
+                return true;
+        }
+        return false;
+}
```
Comment (engelmi) [inline src/libhirte/common/cfg.c:149]:
```suggestion
bool is_file_name_ending_with_conf(const struct dirent *entry) {
        return strstr(entry->d_name, ".conf") != NULL;
}
```
Sorry for being nit-picky, but could you change it like this? This way its a bit more concise.

**Review by engelmi — APPROVED** [review]
LGTM, great work! 
Just one little nit-picky detail that would be great to change, in my point of view.

File: src/libhirte/common/cfg.c (hunk 27 above)
Comment (mwperina) [reply src/libhirte/common/cfg.c:149]: I think the method needs to return `int` to be usable as a filter in scandir:

```
int scandir(const char *restrict dirp,
                   struct dirent ***restrict namelist,
                   int (*filter)(const struct dirent *),
                   int (*compar)(const struct dirent **,
                                 const struct dirent **));
```

Comment (engelmi) [reply src/libhirte/common/cfg.c:149]: You are right. Haven't seen the (changed?) usage of this function.

File: hirte.spec.in (hunk 6 above)
Comment (ArtiomDivak) [reply hirte.spec.in]: Can we close that conversation?
